### PR TITLE
[v6r19] Add new fields in TransformationMonitor, fix ReplicateAndRegister

### PR DESCRIPTION
**Pull request by phicharp**
BEGINRELEASENOTES
* TransformationSystem
CHANGE: add possibility to get extension-specific tasks and files statuses in TransformationMonitor web application

* RMS
NEW: add option --ListJobs to dirac-rms-request to list the jobs for a set of requests

* DMS
FIX: in DMSHelpers, fix case when no site is found for an SE
FIX: in ReplicateAndRegister, don't try and get SE metadata is replica is inactive

ENDRELEASENOTES

### CODE REVIEW

**Comment by phicharp** [comment]
@zmathe : Zoltan, is it sufficient to add these parameters in the method `getTransformationSummaryWeb` for them to be included in the web display? I suspect there is another place where the columns are defined, right?

**Comment by zmathe** [comment]
The new fields have to be added also to WebApp. I will modify later today.

**Comment by phicharp** [comment]
OK, but can these fields also be taken from the CS or does it need to be hardcoded, which is not very nice...

**Comment by zmathe** [comment]
It has to be hardcoded: https://github.com/DIRACGrid/WebAppDIRAC/blob/integration/WebApp/static/DIRAC/TransformationMonitor/classes/TransformationMonitor.js#L40

**Comment by phicharp** [comment]
Too bad... However the statuses 'NotProcessed' and 'Removed' are LHCb-specific, therefore should be in LHCbDIRAC...

**Comment by zmathe** [comment]
The dataFields can be overwritten in LHCbWeb

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 1]
```diff
@@ -29,203 +29,208 @@
...
+    allReplicas = dataManager.getReplicas(opFile.LFN, getUrl=False)
     if allReplicas['OK']:
-      allReplicas = allReplicas['Value']['Successful'].get( opFile.LFN, {} )
+      allReplicas = allReplicas['Value']['Successful'].get(opFile.LFN, {})
       if not allReplicas:
-        ret['NoReplicas'].append( None )
+        ret['NoReplicas'].append(None)
         noReplicas = True
       else:
-        # We try inactive replicas to see if maybe the file doesn't exist at all
```
Comment (phicharp) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: here is the change

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 2]
```diff
@@ -29,203 +29,208 @@
...
     if fcChecksum:
-      if hexAdlerToInt( fcChecksum ) != False:
+      if hexAdlerToInt(fcChecksum) is not False:
         opFile.Checksum = fcChecksum
-        opFile.ChecksumType = fcMetadata['Value']['Successful'][opFile.LFN].get( 'ChecksumType', 'Adler32' )
+        opFile.ChecksumType = fcMetadata['Value']['Successful'][opFile.LFN].get('ChecksumType', 'Adler32')
       else:
         opFile.Checksum = None
 
+  # If no replica was found, return what we collected as information
```
Comment (phicharp) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: ... and return if no replica is found

**Comment by chaen** [comment]
review ok

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 3]
```diff
@@ -239,42 +244,41 @@ def _addMetadataToFiles( self, toSchedule ):
...
-      return S_OK( "%s targets are banned for writing" % ",".join( bannedTargets['Value'] ) )
+      return S_OK("%s targets are banned for writing" % ",".join(bannedTargets['Value']))
 
     # Can continue now
-    self.log.verbose( "No targets banned for writing" )
+    self.log.verbose("No targets banned for writing")
 
     toSchedule = {}
 
+    errors = {}
```
Comment (andresailer) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]:
```
from collections import defaultdict
errors = defaultdict(int)
```

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 4]
```diff
@@ -283,218 +287,238 @@ def ftsTransfer( self ):
...
-          toSchedule[opFile.LFN] = [ opFile, validReplicas, validTargets ]
+          toSchedule[opFile.LFN] = [opFile, validReplicas, validTargets]
       else:
-        gMonitor.addMark( "FTSScheduleFail" )
+        gMonitor.addMark("FTSScheduleFail")
         if noMetaReplicas:
-          self.log.warn( "unable to schedule '%s', couldn't get metadata at %s" % ( opFile.LFN, ','.join( noMetaReplicas ) ) )
-          opFile.Error = "Couldn't get metadata"
+          err = "Couldn't get metadata"
+          errors[err] = errors.setdefault(err, 0) + 1
```
Comment (andresailer) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: when using defaultdict
```
errors[err] += 1
```

Comment (phicharp) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: Yes, thanks for reminding me, this is a great feature! I'll do it

**Comment by phicharp** [comment]
Please review... hopefully that's it for this week!

**Comment by phicharp** [comment]
Diffs: https://github.com/DIRACGrid/DIRAC/pull/3573/files?w=1

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 5]
```diff
@@ -239,262 +245,318 @@ def _addMetadataToFiles( self, toSchedule ):
...
 
+    delayExecution = 0
+    errors = defaultdict(int)
     for opFile in self.getWaitingFilesList():
       opFile.Error = ''
-      gMonitor.addMark( "FTSScheduleAtt" )
+      gMonitor.addMark("FTSScheduleAtt")
       # # check replicas
-      replicas = self._filterReplicas( opFile )
+      replicas = self._filterReplicas(opFile)
```
Comment (atsareg) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: Code for evaluating filtered replicas is similar in ftsTransfer and in dmTransfer. Might be worth making it a common utility

File: RequestManagementSystem/Client/ReqClient.py [hunk 6]
```diff
@@ -402,125 +406,138 @@ def resetFailedRequest( self, requestID, allR = False ):
...
     key += ': '
   blanks = offset * ' '
-  if mainItem and type( mainItem ) == type( {} ):
-    output += "%s%s%s\n" % ( blanks, key, '{' ) if blanks or key else ''
-    for key in sorted( mainItem ):
-      prettyPrint( mainItem[key], key = key, offset = offset )
-    output += "%s%s\n" % ( blanks, '}' ) if blanks else ''
-  elif mainItem and type( mainItem ) == type( [] ) or type( mainItem ) == type( tuple() ):
-    output += "%s%s%s\n" % ( blanks, key, '[' if type( mainItem ) == type( [] ) else '(' )
+  if mainItem and isinstance(mainItem, type({})):
```
Comment (atsareg) [inline RequestManagementSystem/Client/ReqClient.py]: There are multiple replacements for isinstance() which is good but better to make it properly, I mean:

   isinstance(mainItem, type({})) -> isinstance(mainItem, dict)

And similar for other replacements in this file

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py (hunk 5 above)
Comment (phicharp) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: It is the same method that is called...

File: RequestManagementSystem/Client/ReqClient.py (hunk 6 above)
Comment (phicharp) [reply RequestManagementSystem/Client/ReqClient.py]: Ah! This is done automatically by autopep8! I though it was doing it properly, probably though only if the original code is `if type(x) == types.dict`.
I will fix them by hand

**Pull request merged**